src/gui/widgets/LinkedModelGroupLayout.cpp

```diff
@@ -0,0 +1,303 @@
+** A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+** OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+** SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+** LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+** DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+** THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+** (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+** OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE."
+**
+** $QT_END_LICENSE$
+**
+****************************************************************************/
+
+
+#include "LinkedModelGroupLayout.h"
+
+#include <QWidget>
+#include <QLayoutItem>
+#include <QLineEdit>
+#include <QRect>
+#include <QString>
+
+// TODO: rename: ModelGroupLayout
+
+constexpr const int LinkedModelGroupLayout::m_minWidth;
+
+LinkedModelGroupLayout::LinkedModelGroupLayout(QWidget *parent, int margin, int hSpacing, int vSpacing)
+	: QLayout(parent), m_hSpace(hSpacing), m_vSpace(vSpacing),
+	  m_searchBar(new QLineEdit(parent))
+{
+	setContentsMargins(margin, margin, margin, margin);
+	m_searchBar->setPlaceholderText("filter");
+	connect(m_searchBar, SIGNAL(textChanged(const QString&)),
+		this, SLOT(onTextChanged(const QString& )));
+	addWidget(m_searchBar);
+	m_searchBar->setHidden(true); // nothing to filter yet
+}
+
+LinkedModelGroupLayout::~LinkedModelGroupLayout()
+{
+	QLayoutItem *item;
+	while ((item = takeAt(0))) { delete item; }
+}
+
+void LinkedModelGroupLayout::onTextChanged(const QString&)
+{
+	invalidate();
+	update();
+}
+
+void LinkedModelGroupLayout::addItem(QLayoutItem *item)
+{
+	QWidget* widget = item->widget();
+	const QString str = widget ? widget->objectName() : QString("unnamed");
+	m_itemMap.insert(str, item);
+	invalidate();
+}
+
+int LinkedModelGroupLayout::horizontalSpacing() const
+{
+	if (m_hSpace >= 0) { return m_hSpace; }
+	else
+	{
+		return smartSpacing(QStyle::PM_LayoutHorizontalSpacing);
+	}
+}
+
+int LinkedModelGroupLayout::verticalSpacing() const
+{
+	if (m_vSpace >= 0) { return m_vSpace; }
+	else
+	{
+		return smartSpacing(QStyle::PM_LayoutVerticalSpacing);
+	}
+}
+
+int LinkedModelGroupLayout::count() const
+{
+	return m_itemMap.size() - 1;
+}
+
+QMap<QString, QLayoutItem*>::const_iterator
+LinkedModelGroupLayout::pairAt(int index) const
+{
+	if (index < 0)
+		return m_itemMap.cend();
+	QMap<QString, QLayoutItem*>::const_iterator itr = m_itemMap.cbegin();
+	++itr; // skip search bar
+	while (index-->0 && itr != m_itemMap.cend()) { ++itr; }
+	return itr;
+}
+
+// linear time :-(
+QLayoutItem *LinkedModelGroupLayout::itemAt(int index) const
+{
+	auto itr = pairAt(index);
+	return itr == m_itemMap.end() ? nullptr : itr.value();
+}
+
+QLayoutItem *LinkedModelGroupLayout::itemByString(const QString &key) const
+{
+	auto itr = m_itemMap.find(key);
+	return (itr == m_itemMap.end()) ? nullptr : *itr;
```

Comment:
Why is this different (has `()`s) than the `itemAt` return?

Reply:
Good question. I now added the `()` in `itemAt`.